refactor: replace recipes with commands

abeatrix opened the pull request:
The main goal of this PR is to deprecate recipes and migrates the popular recipes into commands, a new alternative to `recipes`.

See Test Plan on how to enable it for testing.

> Note: The Custom Commands feature is still in the exploring stage and has not been finalized. Use at your own risk.

This PR includes:
- 3 new default commands created with a JSON file:
  - Document Code
  - Explain Code
  - Generate Unit Tests
- Added e2e test for Custom Commands
- renamed the `my-cody` directory to `custom-prompts`
- rename custom recipes to custom commands
- refactored custom recipes files and removed unused methods to custom commands
- renamed and clean up methods for custom commands
- remove redundant debug output from LocalAppDetector
- new command menu in command palette, with item to start inline chat
- <img width="906" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/3db7be03-5b54-4107-85be-af58acd973e6">
- command menu can be invoked from icon in doc context menu, or shortcut `cmd + shift + v`
- <img width="299" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/55fa554b-65a4-48e2-9a10-2f3d1764803b">
- new custom commands config menu for managing custom commands config files
- <img width="1009" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/769b1ae3-4321-4891-b161-95c9e355371d">
- added code lenses for Cody that trigger the new command menu in text files
- <img width="1467" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/05642a09-035c-4465-be32-bb101b3ea293">
- New slash commands panel invokable from chat box:
- <img width="833" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/712af66a-795e-425d-ac16-c8ddff9a625e">
- Moved cody items up in context menu
- <img width="941" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/58d4cec9-e307-46b6-ad0f-0c870ddeacf4">
- New cleaner UI with Recipes tab removed
- <img width="2002" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/e7c1a849-2f07-485d-b5af-35aa618e7c40">



## Test plan

<!-- Required. See https://docs.sourcegraph.com/dev/background-information/testing_principles. -->

added e2e test for custom commands and updated current tests to work with new default commands

toolmantim commented on vscode/src/editor/EditorCodeLenses.ts:
```diff
@@ -0,0 +1,230 @@
+    endLineText.length > 0 && !!endLineText.match(new RegExp(`^\\s{${startLineText.length}}.*$`))
+
+const editorCodeLenses = {
+    cody: { title: '$(cody-logo) Cody', command: 'cody.editor.codelens.click', tooltip: 'Open recipes menu' },
+    inline: { title: 'Inline', command: 'cody.editor.codelens.click', tooltip: 'Ask Cody inline' },
```
In line with what we went through today, should we change this now to a single `$(cody-logo) Cody'` code lens action that opens a single quickpick [like this](https://www.figma.com/file/9112BsKsJc1BpO2j8XYwLL/%F0%9F%A4%96-Cody-VS-Code-%5BMain%5D?type=design&node-id=4313-152677&mode=design&t=uwT81rBv1jQFm5yt-4)?

![image](https://github.com/sourcegraph/cody/assets/153/6ca9313e-3aeb-4c7f-8e3b-af2af054b63d)

Feels weird to just have the word "Inline" sitting there…

toolmantim commented on vscode/webviews/Recipes.tsx:
```diff
@@ -2,6 +2,8 @@ import { VSCodeButton } from '@vscode/webview-ui-toolkit/react'
 
```
I was going to suggest some cleanups to this based on the changes I had in #374, but it's all going to get `rm`'d and switched to a quickpick yeah?

toolmantim commented on vscode/src/custom-recipes/CustomRecipesMenus.ts:
```diff
@@ -44,11 +42,11 @@ export async function createNewPrompt(promptName?: string): Promise<CodyPrompt |
     const minPromptLength = 3
```
Any reason it can't be one or two words? If someone wanted to create a "Transmogrify" or "🔥" recipe, who are we to stop them…

toolmantim commented on vscode/src/custom-recipes/helper.ts:
```diff
@@ -85,3 +87,17 @@ Output of \`{command}\` command:
 \`\`\``
+
+export const lastUsedRecipesSeperator: [string, CodyPrompt][] = [
+    ['seperator', { prompt: 'seperator', type: 'last used' }],
```
VS Code style:

```suggestion
    ['seperator', { prompt: 'seperator', type: 'recently used' }],
```

toolmantim commented on vscode/src/custom-recipes/MyPromptController.ts:
```diff
@@ -0,0 +1,375 @@
+                          }
+                }) as vscode.QuickPickItem[]
+            const seperator: vscode.QuickPickItem = { kind: -1, label: 'action' }
+            const addOption: vscode.QuickPickItem = { label: 'Create a New User Recipe', alwaysShow: true }
```
```suggestion
            const addOption: vscode.QuickPickItem = { label: 'New Custom Recipe...', alwaysShow: true }
```

toolmantim reviewed (commented):
I didn't have a chance to run it locally, but reviewed as much of the UI copy as I could

philipp-spiess commented on vscode/src/custom-recipes/CustomRecipesMenus.ts:
```diff
@@ -194,10 +143,12 @@ export async function showRecipeTypeQuickPick(
     const userItem = {
         label: 'User',
         detail: 'User Recipes are accessible only to you across Workspaces',
```
The casing here looks incorrect. This is neither Title Case (used in some VS Code dialogs) nor regular english casing.

It should either be 

```suggestion
        detail: 'User recipes are accessible only to you across workspaces',
```

or when it's shown in the command palette 

```suggestion
        detail: 'User Recipes Are Accessible Only to You Across Workspaces',
```

abeatrix replied:
Ah I was thinking [Workspace](https://code.visualstudio.com/docs/editor/multi-root-workspaces#_workspace-file) and _User Recipes_ are the feature / product names, and therefore are considered as [proper nouns](https://www.grammarly.com/blog/proper-nouns/) which should always be capitalized in a (non-title) sentence. Do you think those rules apply in this case? 👀

philipp-spiess replied:
Interesting. I'll refer to @toolmantim then, he's the native speaker 🙈

philipp-spiess commented on vscode/src/custom-recipes/CustomRecipesMenus.ts:
```diff
@@ -194,10 +143,12 @@ export async function showRecipeTypeQuickPick(
     }
     const workspaceItem = {
         label: 'Workspace',
         detail: 'Workspace Recipes are available to all users in your current repository',
```
Same comment about casing as above

philipp-spiess commented on vscode/src/editor/EditorCodeLenses.ts:
```diff
@@ -0,0 +1,230 @@
+    /**
+     * Get the line where the function ends
+     */
+    private _getFunctionEndLine(startLine: number, lineCount: number, document: vscode.TextDocument): number {
+        if (!document) {
+            return 0
+        }
+        const startLineText = document.lineAt(startLine).text
+        const isStartLineStartedWithWord = startsWithWord(startLineText)
+        // Iterate forwards util we find the line that:
+        // - starts with a } or
+        // - is start of a new function
+        for (let i = startLine + 1; i < lineCount; i++) {
+            const text = document.lineAt(i).text
+            // If the start line text does not start with space or tab
+            // then the end line should start with word without space or tab
+            if (isStartLineStartedWithWord) {
+                if (startsWithWord(text)) {
+                    return i - 1
+                }
+            } else {
+                // if start line end with opening curly bracket and end line starts with closing curly bracket
+                const isStartLineEndedWithOpeningCurlyBracket = startLineText.match(/^\s*{$/)
+                const isEndLineStartedWithClosingCurlyBracket = text.match(/^\s*}$/)
+                if (isStartLineEndedWithOpeningCurlyBracket && isEndLineStartedWithClosingCurlyBracket) {
+                    return i + 1
+                }
+                // if line is the start of a new function
+                const isStartOfFunction = checkIsStartOfFunction(text)
+                if (isStartOfFunction && text.length > 0) {
+                    return i - 1
+                }
+                if (checkHasSameNumberOfSpacesAsStartLine(startLineText, text)) {
+                    return i - 1
+                }
+            }
+        }
+        return lineCount
+    }
```
Can we extract this into a standalone helper and write tests for it? Also this won't work well for non C-like languages, what's the idea there? E.g. Python (which is _very_ prominent)

philipp-spiess commented on vscode/src/editor/EditorCodeLenses.ts:
```diff
@@ -0,0 +1,230 @@
+ * Use regex to check if the line starts with a word character
+ * or if it is a functions or class
+ */
+const checkIsStartOfFunction = (text: string): boolean => {
+    const isEmptyLine = /^\s*$/.test(text)
+    const isArrowFunction = /=.*=>.*$/.test(text)
+    const isNotFunction =
+        !!text.match(/^\s*(if|for|while|switch|case|return|try|catch).*$/) ||
+        (!!text.match(/.*=.*({)?$/) && !isArrowFunction)
+    const isFunction = !!text.match(/^(\s*)?\w.*{$/) && !isNotFunction
+    const isPythonFunction = /^\s*(async\s*)?def\s/.test(text)
+    const isClass = /^(export\s*)?class\s/.test(text)
+    const isComment = /^\s*\/(\/|\*)?/.test(text)
+    return !isEmptyLine && !isComment && (isFunction || isPythonFunction || isClass || isArrowFunction)
```
Same comment as above: Can we extract this and write tests for it? 

Maybe we can reuse this for autocomplete too in some form or another, but it's really unclear how this works and what limitations this has. What are the supported languages etc.?

philipp-spiess commented on vscode/test/e2e/custom-recipes.test.ts:
```diff
@@ -15,11 +15,18 @@ test('open the custom recipes in sidebar', async ({ page, sidebar }) => {
+    await page.keyboard.type('this is a test')
+    await page.keyboard.press('Enter')
+    await page.keyboard.press('Enter')
+    // await sidebar.locator('vscode-button').filter({ hasText: recipeName }).click()
```
```suggestion
```

abeatrix commented:
@toolmantim i still need to fix the tests and update the PR description, but other than that, can you check if the rest of the build looks good to you? I've removed the `Recipes` tab and all the old `recipes`, and replaced the recipes actions with commands.

toolmantim commented:
@abeatrix I just pushed up some copy cleanups, but there'e still lots that isn't working for me (like executing and commands doesn't seem to do anything… and the "New Custom Command…" doesn't seem to trigger the multi-step quickpick like it used to) 🤔 Let's sync to go through it together?

abeatrix commented:
@toolmantim i just pushed some updates so that you won't need to set the feature flag to get those to work. I've also finished with the clean up and updated the PR so it's good to test! Happy to sync up later if you want.

toolmantim commented:
It took a lot of fighting, but I got it looking pretty good across all themes, including high contrast! I also added a proper focus outline to the input too.

https://github.com/sourcegraph/cody/assets/153/0cec9572-7206-4540-b1f3-574d969a2a99

A few outstanding things I couldn't fix @Beatrix:
1. The items are still in reverse alphabetical order (Z-A) but should be A-Z
2. I feel like escape should clear the input, making it feel as fast as the standard quickpick. It's a bit odd for a textarea but I think it makes more sense.
3. The command chooser should disappear when the input loses focus (and reappear when it does, if the input has a slash command in it). A few times I've accidentally deleted code in the editor mis-understanding the focus/cursor.
